### PR TITLE
WIP #81 renaming test kernels to follow convention

### LINKED ISSUE #81: Test kernels do not obey module naming convention
While adding test compilation to #60 I've realised that the vast majority of the Dynamo test kernels do not have the correct module/file name. The module containing a kernel, `my_kernel`, should be named `my_kernel_mod` and be in file `my_kernel_mod.f90`. However, the majority of kernel filenames are missing the `_mod`. This causes problems for test compilation because the compiler cannot find the `.mod` file for the module.

**Info**

The test kernels are in https://github.com/stfc/PSyclone/tree/master/src/psyclone/tests/test_files/dynamo0p3 directory and `kernels` subdirectory within that. The general name pattern to look for kernels is `testkern*_<>_mod.f90` (or `F90`). If there is no `mod` in the name of the file and the module, it should be added in.

### PR DESCRIPTION
**Pull request by mo-lottieturner**
closes #81

### CODE REVIEW

**Comment by mo-lottieturner** [comment]
- [ ] testkern_short_name.F90 (used in dynamo0p3_test.py);
- [ ] xyz (an empty file);
- [ ] testkern_qr.F90 (./dynamo0p3_quadrature_test.py);
- [ ] testkern_no_datatype.F90 (./dynamo0p3_test.py);
- [ ] testkern_invalid_fortran.F90 (./dynamo0p3_test.py);
- [ ] simple.f90 (./dynamo0p3_stubgen_test.py);
- [ ] simple_with_reduction.f90 (domain/lfric/lfric_scalar_stubgen_test.py).
- [ ] kernels/dead_end/no_really/testkern_qr.F90 in src/psyclone/tests/test_files/dynamo0p3 directory,
- [ ] testkern_qr.F90 in examples/line_length directory.

**Comment by mo-lottieturner** [comment]
> Just checking - is this one ready for review?

@arporter I think so! `testkern_qr.F90` in the main `dynamo0p3` directory is meant to not follow convention (the test it is attached to is for files that don't follow the convention). I've left `examples/line_length/testkern_qr.F90` alone for the same reason, though I have updated it so the code inside matches the code in the main `testkern_qr.F90` file. `dead_end/no_really/testkern_qr.F90` doesn't seem to be used in any tests so I _have_ changed that name to follow convention

**Comment by arporter** [comment]
> `testkern_qr.F90` in the main `dynamo0p3` directory is meant to not follow convention (the test it is attached to is for files that don't follow the convention)

Just to check I've got this right, the one test that uses this file is `dynamo0p3_quadrature_test.py::test_dynkern_setup`? As far as I can tell, the naming of the file is not a part of this test so I think `testkern_qr.F90` should be renamed.

File: examples/line_length/testkern_qr.F90 [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ! -----------------------------------------------------------------------------
 ! BSD 3-Clause License
 !
-! Copyright (c) 2017-2021, Science and Technology Facilities Council
+! Copyright (c) 2017-2023, Science and Technology Facilities Council
```
Comment (arporter) [inline examples/line_length/testkern_qr.F90:4]: As discussed in the comments, I think this file needs renaming too.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_qr.F90 [hunk 2]
```diff
@@ -30,7 +30,7 @@
 ! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_qr.F90:31]: As discussed in the comments, please could you also rename this one?

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for this Lottie.
Everything looks good although I think the two files you deliberately didn't rename should be (unless I've missed something).
Tests and examples are all fine with compilation.
Updated User Guide builds fine.
Issue number is not mentioned anywhere.

**Comment by mo-lottieturner** [comment]
@arporter found the snippet that made me think testkern_qr didn't need to be changed - from `doc/user_guide/psyclone_kern.rst`. Thoughts?

```
``testkern_invalid_fortran_mod.f90``, ``testkern_no_datatype_mod.f90``,
``testkern_short_name_mod.f90`` and ``testkern_qr.F90`` are designed to be
invalid for PSyclone stub generation testing purposes and should produce
appropriate errors. Two examples are below::

[...]

As noted above, if the LFRic API naming convention for module and type
names is not followed, the stub generator will return with an error
message. For example::

    > psyclone-kern -gen stub tests/test_files/dynamo0p3/testkern_qr.F90
    Error: "Parse Error: Error, module name 'testkern_qr' does not have
    '_mod' as an extension. This convention is assumed."
```

**Comment by mo-lottieturner** [comment]
changed `testkern_qr` to `testkern_wrong_file_name` as `testkern_wrong_mod_name` already in use

just the sphinx-link-check failing, the link its saying is broken works fine, going to merge the recent master changes onto the branch and then go digging

File: src/psyclone/tests/alggen_test.py [hunk 3]
```diff
@@ -162,7 +162,7 @@ def test_single_function_invoke_qr():
                                    "1.1.0_single_invoke_xyoz_qr.f90"),
                       api="dynamo0.3")
     gen = str(alg).lower()
-    assert "use testkern_qr" not in gen
+    assert "use testkern_wrong_file_name" not in gen
```
Comment (arporter) [inline src/psyclone/tests/alggen_test.py:165]: I think there's some confusion here. We want this file to *only* be used to illustrate what happens if a file has the wrong name. For everything else we should use a valid kernel with a valid filename. (I thought PSyclone would reject such a filename anyway but apparently we don't :-( That's not something for this PR though.) Therefore, perhaps we need a `testkern_qr_mod` file and `testkern_wrong_file_name` too?

**Review by arporter — CHANGES_REQUESTED** [review]
Hi @mo-lottieturner, apologies for the confusion but what I meant was that the file with the wrong name should only be used to illustrate what happens when the name is wrong. Everywhere else we should always be using files with valid names. You may be best reverting your most recent changes and then renaming `testkern_qr` so that it has a valid filename. You just need one new file with an invalid name that can be used in the user guide example (if that's the only place we demonstrate this. Perhaps there's a test or two too?)

**Review by arporter — CHANGES_REQUESTED** [review]
All looks good now, thanks very much @mo-lottieturner.
Ref. guide builds fine.
I was all set to merge but I checked the compilation tests and they fail :-( :100: 

    =========================== short test summary info ============================
    FAILED dynamo0p3_quadrature_test.py::test_field_xyoz - AssertionError: assert False
    FAILED dynamo0p3_quadrature_test.py::test_field_qr_deref - AssertionError: assert False
    FAILED dynamo0p3_basis_test.py::test_two_qr_same_shape - AssertionError: assert False
    FAILED dynamo0p3_basis_test.py::test_two_identical_qr - AssertionError: assert False
    FAILED dynamo0p3_basis_test.py::test_two_qr_different_shapes - AssertionError: assert False
    FAILED dynamo0p3_basis_test.py::test_qr_plus_eval - AssertionError: assert False
    FAILED psyGen_test.py::test_codedkern_module_inline_kernel_in_multiple_invokes - AssertionError: assert False
    FAILED dynamo0p3_basis_test.py::test_2eval_1qr_2fs - AssertionError: assert False
    FAILED domain/lfric/arg_ordering_test.py::test_kerncallarglist_quad_rule_error[False] - AssertionError: assert False
    FAILED domain/lfric/arg_ordering_test.py::test_kerncallarglist_quad_rule_error[True] - AssertionError: assert False
    FAILED psyGen_test.py::test_named_multi_invokes - AssertionError: assert False
    FAILED dynamo0p3_test.py::test_multikernel_invoke_qr - AssertionError: assert False
    FAILED dynamo0p3_test.py::test_kerncallarglist_quad_rule_error[False] - AssertionError: assert False
    FAILED dynamo0p3_test.py::test_kerncallarglist_quad_rule_error[True] - AssertionError: assert False
    FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_check_seq_colours_omp_do[False-False] - AssertionError: assert False
    FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_check_seq_colours_omp_do[False-True] - AssertionError: assert False
    FAILED domain/lfric/transformations/lfric_extract_test.py::test_node_list_error -AssertionError: assert False
    FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_check_seq_colours_omp_do[True-False] - AssertionError: assert False
    FAILED domain/lfric/transformations/dynamo0p3_transformations_test.py::test_check_seq_colours_omp_do[True-True] - AssertionError: assert False
    =========== 19 failed, 5291 passed, 60 xfailed in 214.92s (0:03:34) ============

(Use `pytest --compile my-test-file` to reproduce this.)

**Comment by mo-lottieturner** [comment]
should be working now!

**Comment by TeranIvy** [comment]
I took the liberty of updating the branch with the master and marking it as "ready for review". @mo-lottieturner and @arporter, I hope you won't mind but please tell me off if you do!

**Review by arporter — APPROVED** [review]
Change is good. Integration tests are green.
Will proceed to merge.

**Pull request merged**